tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+    par_base = """
+    PSR                             1023+0038
+    TRACK                                  -3
+    EPHEM                               DE421
+    CLOCK                        TT(BIPM2019)
+    START              54906.7327557001113774
+    FINISH             56443.7950227385911112
+    DILATEFREQ                              N
+    DMDATA                                0.0
+    NTOA                               4970.0
+    CHI2                   14234223.733031187
+    RAJ                     10:23:47.68719801
+    DECJ                     0:38:40.84551000
+    PMRA                                  0.0
+    PMDEC                                 0.0
+    PX                                    0.0
+    POSEPOCH           54995.0000000000000000
+    F0                   592.4214681620181649 0 
+    F1              -2.4181868552365951721e-15 0
+    PEPOCH             55352.6375960000000001
+    PLANET_SHAPIRO                          N
+    DM                  14.720549454038943568 1 
+    DM1                                   0.0
+    BINARY BT_piecewise
+    PB                  0.1980963253021312476 1 
+    PBDOT              -8.277368114929306e-11 1 
+    A1                    0.34333468063634737 1 
+    A1DOT               1.210496074329345e-13 1 
+    ECC                                   0.0
+    EDOT                                  0.0
+    T0                 55361.5927961792717724 1 
+    OM                                    0.0 0 
+    OMDOT                                 0.0
+    GAMMA                                 0.0
+    TZRMJD             56330.1763862586450231
+    TZRSITE                                 8
+    TZRFRQ                           1547.743
+    """
+    model = get_model(StringIO(par_base))
+    return model
+
+
+@pytest.fixture()
+def make_toas_to_go_with_two_piece_model(build_piecewise_model_with_two_pieces):
+    # makes toas to go with the two non-overlapping, complete coverage model
+    m_piecewise = deepcopy(build_piecewise_model_with_two_pieces)
+    m_piecewise.setup()
+    toas = pint.simulation.make_fake_toas_uniform(
+        m_piecewise.START.value + 1, m_piecewise.FINISH.value - 1, 20, m_piecewise
+    )  # slightly within group edges to make toas unambiguously contained within groups
+    return toas
+
+
+def get_toa_group_indexes(model, toas):
+    # returns array of group indexes associated with each toa. (i.e. which group is each toa in)
+    model.setup()
+    index = model.which_group_is_toa_in(toas)
+    return index
+
+
+def get_number_of_groups(model):
+    # returns number of groups
+    model.setup()
+    number_of_groups = model.get_number_of_groups()
+    print(model.as_parfile())
+    print(model.get_number_of_groups())
+    return number_of_groups
+
+
+@pytest.fixture()
+def build_piecewise_model_with_two_pieces(build_model):
+    # takes the basic model frame and adds 2 non-ovelerlapping pieces to it
+    piecewise_model = deepcopy(build_model)
```

Comment:
Fixtures always give you a fresh-created new object, so there's no need to copy it.